cpu/stm32_common/periph/i2c_1.c

```diff
@@ -146,365 +147,260 @@ int i2c_release(i2c_t dev)
     }
-
-    uint16_t tick = TICK_TIMEOUT;
-
-    I2C_TypeDef *i2c = i2c_config[dev].dev;
-
-    /* Check to see if the bus is busy */
-    while ((i2c->ISR & I2C_ISR_BUSY) && tick--) {}
-    if (!tick) {
-        return -ETIMEDOUT;
+    DEBUG("[i2c] read_bytes: Starting\n");
+    int ret = _start(i2c,
+        (address << 1) |
+        (length << I2C_CR2_NBYTES_Pos) |
+        /* RELOAD is needed because we don't know the full frame */
+        I2C_CR2_RELOAD |
+        I2C_FLAG_READ,
+        flags);
+    if (ret < 0) {
+        return ret;
     }
 
-    if (!(flags & I2C_NOSTART)) {
-        DEBUG("[i2c] read_regs: send start sequence\n");
-        /* send start sequence and slave address */
-        int ret = _start(i2c, address, 1, 0, flags);
+    for (size_t i = 0; i < length; i++) {
+        /* wait for transfer to finish */
+        DEBUG("[i2c] read_bytes: Waiting for DR to be full\n");
+        ret = _wait_isr_set(i2c, I2C_ISR_RXNE, flags);
         if (ret < 0) {
-            _stop(i2c);
             return ret;
         }
+        /* read data from data register */
+        ((uint8_t*)data)[i]= i2c->RXDR;
+        DEBUG("[i2c] read_bytes: DR full, read 0x%02X\n", ((uint8_t*)data)[i]);
     }
-
-    tick = TICK_TIMEOUT;
-    /* wait for ack */
-    while (!(i2c->ISR & I2C_ISR_TXIS) && tick--) {
-        if ((i2c->ISR & ERROR_FLAG) || !tick) {
-            /* end transmission */
-            _stop(i2c);
-            return -ENXIO;
-        }
+    if (flags & I2C_NOSTOP) {
+        /* With NOSTOP, the TCR indicates that the next command is ready */
+        /* TCR is needed because RELOAD is set preventing a NACK on last byte */
+        return _wait_isr_set(i2c, I2C_ISR_TCR, flags);
+    }
+    /* Wait until stop before other commands are sent */
+    ret = _wait_isr_set(i2c, I2C_ISR_STOPF, flags);
+    if (ret < 0) {
+        return ret;
     }
 
-    DEBUG("[i2c] read_regs: Write register to read\n");
-    i2c->TXDR = reg;
-
-    /* send repeated start sequence, read registers and end transmission */
-    DEBUG("[i2c] read_regs: ACK received, send repeated start sequence\n");
-    return i2c_read_bytes(dev, address, data, length, 0);
+    return _wait_for_bus(i2c);
 }
 
+/**
+ * Cannot support continuous writes or fram splitting at this level.  If an
+ * I2C_NOSTOP has been sent it must be followed by a repeated start or stop.
+ */
 int i2c_write_bytes(i2c_t dev, uint16_t address, const void *data,
                     size_t length, uint8_t flags)
 {
     assert(dev < I2C_NUMOF);
-
-    if (flags & I2C_ADDR10) {
-        return -EOPNOTSUPP;
-    }
-
     I2C_TypeDef *i2c = i2c_config[dev].dev;
+    DEBUG("[i2c] write_bytes: Starting\n");
+    return _write(i2c, address, data, length, flags, 0);
+}
 
-    int ret = 0;
+static int _write(I2C_TypeDef *i2c, uint16_t address, const void *data,
+                    size_t length, uint8_t flags, uint32_t cr2_flags)
+{
+    assert(i2c != NULL);
+    assert(length < 256);
```

Comment:
Why the length limitation or why not combine both in a macro?
If length then it is because I would have to do a check to see if it is greater, if it is set the reload bit and keep track of everything sent.  It adds extra steps and I don't know how valuable it would be (even for something like EEPROM, when you are talking about 256 bytes adding a start and stop frame isn't that much overhead).